File: mqtt_flutter_app/lib/sensors/sensor_factory.dart
import 'sensor.dart';
import 'sensor_temperature.dart';
import 'sensor_smoke.dart';

class SensorFactory {
  static Sensor fromJson(Map<String, dynamic> json) {
    final type = SensorType.values.byName(json['sensor_type']);
    final DateTime lastUpdate = DateTime.fromMicrosecondsSinceEpoch(1000 * (json['timestamp'] as num).toInt(), isUtc: true);

    switch (type) {
      case SensorType.temperature:
        return TemperatureSensor(
          id: json['sensor_id'],
          name: json['sensor_name'],
          lastUpdated: lastUpdate,
          temperatureCelsius: json['data']['temperature'],
        );

      case SensorType.smoke:
        return SmokeSensor(
          id: json['sensor_id'],
          name: json['sensor_name'],
          lastUpdated: lastUpdate,
          smokeDetected: json['data']['smoke_detected'],
          obscuration: json['data']['obscuration'],
        );

      default:
        throw UnsupportedError("Unknown sensor type");
    }
  }
}

File: mqtt_esp32_simulator/sensor_factory.py
from sensor_smoke import SmokeSensor
from sensor_temperature import TemperatureSensor

class SensorFactory:
    @staticmethod
    def create(sensor_type, sensor_id, name):
        if sensor_type == "temperature":
            return TemperatureSensor(sensor_id, name)
        elif sensor_type == "smoke":
            return SmokeSensor(sensor_id, name)
        else:
            raise ValueError(f"Unknown sensor type: {sensor_type}")